docs/proposals/kubelet-raw-metrics-api.md

```diff
@@ -0,0 +1,110 @@
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/compute-resource-metrics-api.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Kubelet raw metrics API
+
+Kubelet currently exposes raw container metrics through the `/stats/` endpoint
+that serves raw container stats. However, this endpoint serves individual
+container stats, and applications like heapster, which aggregates metrics across
+the cluster, must repeatedly query for each container. The high QPS combined
+with the potential data size of raw stats puts unnecessary load on the system
+that could be avoided with an aggregate API.
+
+## Requirements:
+- Provide machine level metrics, all pod metrics (in single request), specific
+  pod metrics
+- Ability to authenticate machine & pod metrics independently from each other
+- Eventually support multiple kinds of metrics (e.g. derived/aggregate type)
+- Follow existing API conventions, fully compatible types able to eventually be
+  served by apiserver library
+
+## Design
+
+Kubelet will start providing a versioned API, following convensions of the API
+server. The metrics API will be its own [API group](api-group.md), and may
+eventually be reused at the cluster-level (outside the scope of this proposal).
+
+### Endpoints
+
+All endpoints are GET endpoints, rooted at `/apis/metrics/v1alpha1/`
+
+- `/` - discovery endpoint; type resource list
+- `/nodes` - host metrics; type `[]metrics.Node`
+  - `/nodes/localhost` - When served by kubelet, the only node provided is
+    `localhost`; type metrics.Node
+- `/pods` - All pod metrics across all namespaces; type `[]metrics.Pod`
+- `/namespaces/{namespace}/pods` - All pod metrics within namespace; type
+  `[]metrics.Pod`
+- `/namespaces/{namespace}/pods/{pod}` - metrics for specific pod; type
+  `metrics.Pod`
+- Unsupported paths return status not found (404)
+  - `/namespaces/`
+  - `/namespaces/{namespace}`
+
+Additionally, all endpoints (except root discovery endpoint) support the
+following optional query parameters:
+
+- `start` - start time to return metrics from; type json encoded `time.Time`
+- `end` - end time to return metrics to; type json encoded `time.Time`
+- `count` - maximum number of stats to return in each ContainerMetrics instance;
+  type int
```

Comment:
Ok, I went with step. However, I disagree that the default should be variable based on the window size. I think it will be better to have a default (predictable) default size. For instance, if I make a request and get back stats ever 10 seconds, and then increase my window size by 30 seconds, I'd expect to get 3 more stats back, not get the same number spaced 11 seconds apart.